File: src/main/java/com/example/cqrs/query/dto/AccountTransactionResponse.java
package com.example.cqrs.query.dto;

import com.example.cqrs.command.entity.event.AbstractAccountEvent;
import com.example.cqrs.command.entity.event.enumerate.EventStatus;
import lombok.*;

import java.time.LocalDateTime;

/**
 * 계좌 거래 이력에 대한 Response DTO입니다.
 */
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@NoArgsConstructor(access = AccessLevel.PRIVATE)
@Builder
@Getter
public class AccountTransactionResponse {

    private String accountId;               // 계좌 ID
    private String eventType;               // 이벤트 타입
    private Double amount;                  // 거래 금액
    private LocalDateTime eventDate;        // 이벤트 발생 일시
    private EventStatus status;             // 이벤트 상태
    private EventMetadataResponse metadata; // 이벤트 메타데이터

    // factory method
    public static AccountTransactionResponse from(AbstractAccountEvent event) {
        return AccountTransactionResponse.builder()
                .accountId(event.getAccountId())
                .eventType(event.getClass().getSimpleName())
                .amount(event.getAmount())
                .eventDate(event.getEventDate())
                .metadata(EventMetadataResponse.from(event.getMetadata()))
                .build();
    }

}

File: src/main/java/com/example/cqrs/command/dto/read/AccountTransactionResponse.kt
package com.example.cqrs.command.dto.read

import com.example.cqrs.command.entity.event.enumerate.EventStatus
import java.time.LocalDateTime

data class AccountTransactionResponse(
    val accountId: String,
    val eventType: String,
    val amount: Double,
    val eventDate: LocalDateTime,
    val status: EventStatus,
    val metadata: EventMetadataResponse
) {
    companion object {
        fun from(
            accountId: String,
            eventType: String,
            amount: Double,
            eventDate: LocalDateTime,
            status: EventStatus,
            metadata: EventMetadataResponse
        ): AccountTransactionResponse {
            return AccountTransactionResponse(accountId, eventType, amount, eventDate, status, metadata)
        }
    }
}
